balancer.go

```diff
@@ -446,13 +496,60 @@ func (c *connManager) reconcileAddresses(addrs []resolver.Address, updateFunc fu
 		newAddrs = append(newAddrs, want...)
 	}
 
+	c.connsByAddr = remaining
+	c.doUpdate(newAddrs, toRemove)
+}
+
+func (c *connManager) doUpdate(newAddrs []resolver.Address, toRemove []conn.Conn) {
 	// we make a single call to update connections in batch to create a single
 	// new picker (avoids potential picker churn from making one change at a time)
-	newConns := updateFunc(newAddrs, toRemove)
-	// add newConns to remaining to compute new set of connections
-	for _, c := range newConns {
-		hostPort := c.Address().HostPort
-		remaining[hostPort] = append(remaining[hostPort], c)
+	newConns := c.updateFunc(newAddrs, toRemove)
+	// add newConns to set of connections
+	for _, cn := range newConns {
+		hostPort := cn.Address().HostPort
+		c.connsByAddr[hostPort] = append(c.connsByAddr[hostPort], cn)
 	}
-	c.connsByAddr = remaining
+}
+
+func (c *connManager) recycleConns(connsToRecycle []conn.Conn) {
+	var needToCompact bool
+	for i, cn := range connsToRecycle {
+		addr := cn.Address().HostPort
+		existing := c.connsByAddr[addr]
+		var found bool
+		for i, existingConn := range existing {
+			if existingConn == cn {
+				found = true
+				// remove cn from the slice
+				copy(existing[i:], existing[i+1:])
+				c.connsByAddr[addr] = existing[:len(existing)-1]
+				break
+			}
+		}
+		if !found {
+			// this connection has already been closed/removed
+			connsToRecycle[i] = nil
+			needToCompact = true
+		}
+	}
+	if needToCompact {
+		i := 0
+		for _, cn := range connsToRecycle {
+			if cn != nil {
+				connsToRecycle[i] = cn
+				i++
+			}
+		}
+		if i == 0 {
+			// nothing to actually recycle
+			return
+		}
+		connsToRecycle = connsToRecycle[:i]
```

Comment:
This loop seems like it's basically `slices.DeleteFunc(connsToRecycle, func(conn conn.Conn) bool { return conn == nil })`, but that was introduced in Go 1.21 and we currently only depend on Go 1.19. WDYT, worth a TODO? I generally prefer simple loops over terse one-liners but it's always a little hard to follow the logic in loops that mutate the slice they're iterating over, so the value is more apparent in cases like these.

Reply:
Definitely worth a TODO. There's a PR in flight to update this to use 1.20 (#57), but still not 1.21.